File: solutions/046.permutations/permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
const permute = nums => {
  let res = []

  const backtrack = (temp, visited) => {

    if (visited.every(e => e === 1)) {
      res.push(temp)
      return
    }

    for (let i = 0; i < nums.length; i++) {
      if (visited[i] === 1) continue
      let item = nums[i]

      visited[i] = 1
      backtrack([...temp, item], visited)
      visited[i] = undefined
    }

  }

  backtrack([], Array(nums.length).fill(0))

  return res
}

File: solutions/046.permutations/permutations.py
class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]

        bfs with visited flags
        """
        ret = []

        def bfs(temp, visited=[0]*len(nums)):
            nonlocal ret

            if len(temp) == len(visited):
                ret.append(temp)

            for i, v in enumerate(visited):
                if v:
                    continue

                visited[i] = 1
                bfs(temp + [nums[i]], visited)
                visited[i] = 0

        bfs([])
        return ret
